lib/constable/permalink_type.ex

```diff
@@ -0,0 +1,28 @@
+defmodule Constable.PermalinkType do
+  @behaviour Ecto.Type
+
+  def type, do: :id
+
+  def cast(integer) when is_integer(integer) do
+    {:ok, integer}
+  end
+
+  def cast(string) when is_binary(string) do
```

Comment:
I'm sorry if I'm slow, but I can't seem to understand why the extra type here is necessary. Where is the app casting the `id` as a string with the slug?

Reply:
Yeah, this is a little confusing. Maybe I should do a better job of documenting this somehow in the code itself somehow. I'm open to ideas. 

Say you have an announcement with title `Elixir is awesome` and the id is `23`. Because we have defined the `Phoenix.Param` protocol as `"#{id}-#{slug}"`, the announcement url will look like this, `announcements/23-elixir-is-awesome`. 

When that hits the controllers, we just consider `23-elixir-is-awesome` as the `:id` since the route is `announcements/:id` (even though the whole slug is not the announcement's database id). So when we retrieve or insert any announcements, we are passing `%Announcement{"id" => "23-elixir-is-awesome"}` to ecto. 

Since we have defined the custom type, ecto will cast that id and use this function we're looking at in this comment (which is why we're checking for binary). At that point, it'll split the `"23-elixir-is-awesome"`, and just use the `23` which is the actual database id for getting/updating the database records. 

I am really open to suggestions on how to best tackle this, and if this is not the right solution for slugs, I'm also happy to scrap this altogether. I don't want to introduce confusing code if it's not needed. This was just the best way I could think of that's low effort to add ids. But it is a little bit magical because it's the combination of `Phoenix.Param` and this ecto type that are working in conjunction.

Reply:
This seems totally reasonable after your explanation.

The only other thing you may consider is that you don't need to always look it up by `Repo.get` and you could use a module for doing your lookups. The only benefit I see there is that your associations don't need to know that it's a special type.

Reply:
Hmmm. Something like that could really work. In a previous iteration, I tried doing something like that. In that iteration I didn't have the `id-slugified-title` format. I just had the `slug` so it was a pain when I had to create/update the associated models because they needed access to the id. So I would have to get the announcement by slug, in order to get the id, just so I could add/update a comment. But now, I could add a plug that stripped away the slug in the controllers that needed it, and just have the id there. That also separates the responsibilities better (in my opinion) since the slugs in the `id-slugified-title` are simply a web concern. 

I'll consider that, and it might actually simply this a lot more.

Reply:
I gave this a shot. I'm actually not sure which I prefer. They both seem okay. If you're interested in this, I have separate branch with the alternate implementation which uses a `plug` to de-slugify the slug before hitting the controllers. 

The comparison with this branch is here -> https://github.com/thoughtbot/constable/compare/gv-url-id-slugs...gv-slug-ids-plug

Reply:
@drapergeek I ended up going with a different approach from what I had here with the custom ecto type. I started feeling like this approach coupled a view concern (the url having a slug) with how we were persisting the data in the database. I didn't really like that, so I just went with a plug that removes the slugified part. If you're interested in looking at that, it's on #489. Everything else in that PR is the same as it was here.